storage/client/queries/queries.go

```diff
@@ -401,6 +410,18 @@ const (
 			DECODE(evs.body ->> 'address', 'base64')=preimages.address_data AND
 			preimages.context_identifier = 'oasis-runtime-sdk/address: secp256k1eth' AND
 			preimages.context_version = 0
+		LEFT JOIN chain.address_preimages AS pre1 ON
```

Comment:
Given that we add three JOINs, do you have a sense of the performance impact? I know an automated way of measuring it is still on our backlog. But we can at least look at EXPLAIN ANALYZE before and after this change, probably for the query behind the `/events` endpoint (= without WHERE) which should be the most impacted by this.

Reply:
It seems to double the query time from ~26ms -> ~55ms on production when querying for 1000 events. 

Original
```
 QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=4671.98..22289.90 rows=1000 width=690) (actual time=13.022..26.602 rows=1000 loops=1)
   ->  Nested Loop Left Join  (cost=4671.98..1555995696.47 rows=88318687 width=690) (actual time=13.020..26.504 rows=1000 loops=1)
         Join Filter: (evs.runtime = tokens.runtime)
         ->  Nested Loop Left Join  (cost=4671.70..872804537.87 rows=88318687 width=695) (actual time=13.006..23.045 rows=1000 loops=1)
               ->  Gather Merge  (cost=4671.27..173327792.42 rows=88318687 width=648) (actual time=12.985..21.834 rows=1000 loops=1)
                     Workers Planned: 2
                     Workers Launched: 2
                     ->  Incremental Sort  (cost=3671.25..163132624.50 rows=36799453 width=648) (actual time=0.466..3.214 rows=433 loops=3)
                           Sort Key: evs.round DESC, evs.tx_index, evs.type, ((evs.body)::text)
                           Presorted Key: evs.round
                           Full-sort Groups: 14  Sort Method: quicksort  Average Memory: 42kB  Peak Memory: 42kB
                           Worker 0:  Full-sort Groups: 12  Sort Method: quicksort  Average Memory: 43kB  Peak Memory: 43kB
                           Worker 1:  Full-sort Groups: 13  Sort Method: quicksort  Average Memory: 45kB  Peak Memory: 45kB
                           ->  Parallel Index Scan Backward using ix_runtime_events_round on runtime_events evs  (cost=0.57..159787681.75 rows=36799453 width=648) (actual time=0.042..1.639 rows=456 loops=3)
                                 Index Cond: (runtime = 'emerald'::runtime)
               ->  Index Scan using ix_address_preimages_address_data on address_preimages preimages  (cost=0.43..7.91 rows=1 width=68) (actual time=0.001..0.001 rows=0 loops=1000)
                     Index Cond: (address_data = decode((evs.body ->> 'address'::text), 'base64'::text))
                     Filter: ((context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'::text) AND ((context_version)::integer = 0))
         ->  Index Scan using evm_tokens_pkey on evm_tokens tokens  (cost=0.28..7.72 rows=1 width=65) (actual time=0.001..0.001 rows=0 loops=1000)
               Index Cond: ((runtime = 'emerald'::runtime) AND ((token_address)::text = (preimages.address)::text))
               Filter: (token_type IS NOT NULL)
 Planning Time: 0.582 ms
 Execution Time: 26.803 ms
```

Updated

```
QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=4676.42..46419.76 rows=1000 width=882) (actual time=13.402..51.315 rows=1000 loops=1)
   ->  Nested Loop Left Join  (cost=4676.42..3686726105.17 rows=88318787 width=882) (actual time=13.401..51.200 rows=1000 loops=1)
         Join Filter: (evs.runtime = tokens.runtime)
         ->  Nested Loop Left Join  (cost=4676.14..3003534173.03 rows=88318787 width=887) (actual time=13.389..47.924 rows=1000 loops=1)
               ->  Nested Loop Left Join  (cost=4675.58..2293507443.36 rows=88318787 width=823) (actual time=13.383..46.837 rows=1000 loops=1)
                     ->  Nested Loop Left Join  (cost=4675.03..1583480713.70 rows=88318787 width=759) (actual time=13.351..34.560 rows=1000 loops=1)
                           ->  Nested Loop Left Join  (cost=4674.47..873453984.03 rows=88318787 width=695) (actual time=13.307..21.939 rows=1000 loops=1)
                                 ->  Gather Merge  (cost=4674.04..173450324.08 rows=88318787 width=648) (actual time=13.293..20.881 rows=1000 loops=1)
                                       Workers Planned: 2
                                       Workers Launched: 2
                                       ->  Incremental Sort  (cost=3674.01..163255144.62 rows=36799495 width=648) (actual time=0.565..3.288 rows=435 loops=3)
                                             Sort Key: evs.round DESC, evs.tx_index, evs.type, ((evs.body)::text)
                                             Presorted Key: evs.round
                                             Full-sort Groups: 13  Sort Method: quicksort  Average Memory: 43kB  Peak Memory: 45kB
                                             Worker 0:  Full-sort Groups: 12  Sort Method: quicksort  Average Memory: 43kB  Peak Memory: 43kB
                                             Worker 1:  Full-sort Groups: 13  Sort Method: quicksort  Average Memory: 45kB  Peak Memory: 45kB
                                             ->  Parallel Index Scan Backward using ix_runtime_events_round on runtime_events evs  (cost=0.57..159910197.59 rows=36799495 width=648) (actual time=0.061..1.617 rows=450 loops=3)
                                                   Index Cond: (runtime = 'emerald'::runtime)
                                 ->  Index Scan using ix_address_preimages_address_data on address_preimages preimages  (cost=0.43..7.92 rows=1 width=68) (actual time=0.000..0.001 rows=0 loops=1000)
                                       Index Cond: (address_data = decode((evs.body ->> 'address'::text), 'base64'::text))
                                       Filter: ((context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'::text) AND ((context_version)::integer = 0))
                           ->  Index Scan using address_preimages_pkey on address_preimages pre_from  (cost=0.56..8.04 rows=1 width=111) (actual time=0.012..0.012 rows=0 loops=1000)
                                 Index Cond: ((address)::text = (evs.body ->> 'from'::text))
                                 Filter: ((context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'::text) AND ((context_version)::integer = 0))
                                 Rows Removed by Filter: 1
                     ->  Index Scan using address_preimages_pkey on address_preimages pre_to  (cost=0.56..8.04 rows=1 width=111) (actual time=0.012..0.012 rows=0 loops=1000)
                           Index Cond: ((address)::text = (evs.body ->> 'to'::text))
                           Filter: ((context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'::text) AND ((context_version)::integer = 0))
                           Rows Removed by Filter: 0
               ->  Index Scan using address_preimages_pkey on address_preimages pre_owner  (cost=0.56..8.04 rows=1 width=111) (actual time=0.000..0.000 rows=0 loops=1000)
                     Index Cond: ((address)::text = (evs.body ->> 'owner'::text))
                     Filter: ((context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'::text) AND ((context_version)::integer = 0))
         ->  Index Scan using evm_tokens_pkey on evm_tokens tokens  (cost=0.28..7.72 rows=1 width=65) (actual time=0.001..0.001 rows=0 loops=1000)
               Index Cond: ((runtime = 'emerald'::runtime) AND ((token_address)::text = (preimages.address)::text))
               Filter: (token_type IS NOT NULL)
 Planning Time: 2.574 ms
 Execution Time: 51.606 ms
(37 rows)
```

Reply:
OK great, thank you for checking! That seems well within reason.